File: Philosopher.py
import threading, time, random

class Philosopher(threading.Thread):

  def __init__(self, name, left_chopstick, right_chopstick, color):
    threading.Thread.__init__(self)
    self.stopped = False
    self.name = color + name + '\033[00m'
    self.left_chopstick = left_chopstick
    self.right_chopstick = right_chopstick

  def pickup(self):
    if not self.stopped:
      self.left_chopstick.acquire()
      self.right_chopstick.acquire()

  def putdown(self):
    self.left_chopstick.release()
    self.right_chopstick.release()

  def stop(self):
    self.stopped = True

  def run(self):
    while not self.stopped:
      print(f'{self.name} is waiting.')
      self.pickup()
      if self.stopped: return

      print(f'{self.name} is eating.')
      time.sleep(random.randint(3, 10))
      self.putdown()

      print(f'{self.name} is thinking.')
      if not self.stopped: time.sleep(random.randint(3, 15))

File: golang/Philosopher.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

// Philosopher structure
type Philosopher struct {
	name      string
	leftFork  *Fork
	rightFork *Fork
	stopped   bool
}

func newPhilosopher(name string, left *Fork, right *Fork) *Philosopher {
	p := Philosopher{name: name, leftFork: left, rightFork: right, stopped: false}
	return &p
}

func (p *Philosopher) think() {
	fmt.Println(p.name, " is thinking")
	p.execute()
}

func (p *Philosopher) eat() {
	fmt.Println(p.name, " is eating")
	p.execute()
}

func (p *Philosopher) execute() {
	if !p.stopped {
		rand.Seed(time.Now().UnixNano())
		min := 1000
		max := 4000
		time.Sleep(time.Duration(rand.Intn((max-min+1)+min)) * time.Millisecond)
	}
}

func (p *Philosopher) pickUp() {
	p.leftFork.mux.Lock()
	p.rightFork.mux.Lock()
}

func (p *Philosopher) putDown() {
	p.leftFork.mux.Unlock()
	p.rightFork.mux.Unlock()
}

func (p *Philosopher) waiting() {
	fmt.Println(p.name, " is waiting")
	p.execute()
}

func (p *Philosopher) start(wg *sync.WaitGroup) {
	fmt.Println(p.name, " has taken a seat")
	p.execute()
	for !p.stopped {
		p.think()
		if p.stopped {
			break
		}
		p.pickUp()
		p.eat()
		p.putDown()
	}
	fmt.Println(p.name, " has finished eating")
	wg.Done()
}

func (p *Philosopher) stop() {
	p.stopped = true
}
